baiji/serialization/json.py

```diff
@@ -23,83 +24,140 @@ def _load(f, *args, **kwargs):
+        self.method_list.insert(index, method)
 
-def _dump_args(kwargs):
-    if kwargs.get('primitive', False):
-        kwargs.update(default=_json_encode_primitive)
+    def clear(self):
+        # be defensive if someone forgets to call super pylint: disable=attribute-defined-outside-init
+        self.method_list = []
+
+    def __call__(self, x):
+        from itertools import ifilter
+        try:
+            return next(ifilter(lambda x: x is not None, (f(x) for f in self.method_list)))
+        except StopIteration:
+            return self.default(x)
+
+    def default(self, x):
+        return x
+
+class JSONDecoder(MethodListCaller):
+    def __init__(self):
+        self.register(self.decode_numpy)
+        self.register(self.decode_scipy)
+        self.register(self.decode)
+
+    def decode(self, dct):
+        pass
+
+    def decode_numpy(self, dct):
+        if '__ndarray__' in dct:
+            try:
+                import numpy as np
+            except ImportError:
+                raise ImportError("JSON file contains numpy arrays; install numpy to load it")
+            if 'dtype' in dct:
+                dtype = np.dtype(dct['dtype'])
+            else:
+                dtype = np.float64
+            return np.array(dct['__ndarray__'], dtype=dtype)
+
+    def decode_scipy(self, dct):
+        if '__scipy.sparse.sparsematrix__' in dct:
+            if not 'dtype' in dct and 'shape' in dct and 'data' in dct and 'format' in dct and 'row' in dct and 'col' in dct:
+                return dct
+            try:
+                import numpy as np
+                import scipy.sparse as sp
+            except ImportError:
+                raise ImportError("JSON file contains scipy.sparse arrays; install numpy and scipy to load it")
+            coo = sp.coo_matrix((dct['data'], (dct['row'], dct['col'])), shape=dct['shape'], dtype=np.dtype(dct['dtype']))
+            return coo.asformat(dct['format'])
+
+
+def _load_args(kwargs):
+    if 'object_hook' in kwargs:
+        raise ValueError("Instead of explicitly setting object_hook, subclass JSONDecoder and pass it as decoder")
+    if 'decoder' in kwargs:
+        kwargs['object_hook'] = kwargs['decoder']
+        del kwargs['decoder']
     else:
-        kwargs.update(default=_json_encode)
-    if 'primitive' in kwargs:
-        del kwargs['primitive']
-    kwargs.update(for_json=True)
+        kwargs['object_hook'] = JSONDecoder()
     return kwargs
 
-def _json_decode(dct):
-    '''
-    Handle Custom json decoding and internal double underscore annotated Direct object deserialization
-    '''
-    if '__ndarray__' in dct:
+
+class JSONEncoder(MethodListCaller):
+    def __init__(self, primitive=ENCODE_PRIMITIVES_BY_DEFAULT):
+        self.primitive = primitive
+        self.register(self.encode_numpy)
+        self.register(self.encode_scipy)
+        self.register(self.encode)
+
+    def default(self, x):
+        return None
+
+    def encode(self, obj):
```

Comment:
Indeed, the context cleared things up.
Probably a good thing for this PR that I did not have context, though ;)